Start testing the PSA built-in drivers: hashes

gilles-peskine-arm opened the pull request:
This is one step on the way to https://github.com/Mbed-TLS/mbedtls/issues/7801 — see the issue for the long-term goal.

In this pull request:

* I introduce a new framework to generate test data for cryptographic operations through the PSA API (or not, for that matter), building on the existing work to generate test data for key creation. This is still the beginning, I expect to do more refactoring later when we figure out what our needs are.
* The test data generator only supports hashes for now, and only for a subset of algorithms (the ones that don't rely on any third-party Python library or a specific version of Python or openSSL).
* I implement some basic tests for hashes: mostly nominal behavior, but also calling a finish function with a larger output buffer, which is the equivalent for hashes of the problem we ran into with mac. The hash functions behave correctly.

Follow-ups:

* (high priority) Write test functions for mac operations and extend the test data generator to mac.
* Support all hashes algorithms (with cryptodome?).
* Extend the framework to other kinds of operations.
* Extend the test coverage to be good coverage for our implementation.
* Extend the test coverage and the tooling so that these tests can also run with a third-party driver.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** no (test only)
- [ ] **backport** TODO (at least the test framework stuff)
- [x] **tests** yes

tgonzalezorlandoarm reviewed (commented):
Minor questions and comments

gilles-peskine-arm commented on scripts/mbedtls_dev/crypto_data_tests.py:
```diff
@@ -0,0 +1,123 @@
+# limitations under the License.
+
+import hashlib
+from typing import Callable, Dict, Iterator, List, Optional #pylint: disable=unused-import
```
We have type annotations that are compatible with Python 3.5 syntax, so variable definitions need type annotations in comments (`foo = []; type: List[str]`). Pylint doesn't parse those annotations, so it thinks they're unused. Mypy parses the annotations and it needs the type names (or `typing` itself if the type names are fully qualified) to be imported.

tgonzalezorlandoarm replied:
Understood, thanks

gilles-peskine-arm commented on scripts/mbedtls_dev/psa_information.py:
```diff
@@ -0,0 +1,151 @@
+# failure.
+def read_implemented_dependencies(filename: str) -> FrozenSet[str]:
+    return frozenset(symbol
+                     for line in open(filename)
```
It's best practice to use a `with` statement, but not outright wrong to leave it out: Python closes the file when the file object goes out of scope.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_low_hash.function:
```diff
@@ -0,0 +1,225 @@
+    mbedtls_free(output);
+    output = NULL;
+
+#if 0
```
Dead code I forgot to remove.

tgonzalezorlandoarm replied:
Just in case: there is another if 0 in this file.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_low_hash.function:
```diff
@@ -0,0 +1,225 @@
+    output = NULL;
+#endif
+
+exit:
```
Test assertion macros (`TEST_ASSERT`, etc.) do `goto exit` if the test assertion fails. Most test functions have cleanup code: we clean resources even if the test fails, so that memory leak frameworks (ASan, Valgrind) don't signal spurious extra leaks and so that low-memory devices don't run out of memory.

tgonzalezorlandoarm replied:
I missed the step of the TEST_ASSERT going to exit. Thanks!

tgonzalezorlandoarm reviewed (changes requested):
I think after solving those comments I can approve this PR

gilles-peskine-arm commented on scripts/mbedtls_dev/psa_information.py:
```diff
@@ -0,0 +1,151 @@
+    return frozenset(symbol
+                     for line in open(filename)
+                     for symbol in re.findall(r'\bPSA_WANT_\w+\b', line))
+_implemented_dependencies = None #type: Optional[FrozenSet[str]] #pylint: disable=invalid-name
```
This bit of code was just moved from another file, and it's changing, so I'd rather avoid cosmetic improvements.

gilles-peskine-arm commented:
There's a conflict in the code that was moved from `generate_psa_tests.py` to `psa_information.py`. Since that part of the code is under active development (driver-only ECC, FFDH driver support), another conflict is possible, so I'll wait to rebase or merge until this pull request has been approved.

gilles-peskine-arm commented on scripts/mbedtls_dev/crypto_data_tests.py:
```diff
@@ -0,0 +1,123 @@
+        'PSA_ALG_SHA3_384': None, #lambda data: hashlib.sha3_384(data).hexdigest(),
+        'PSA_ALG_SHA3_512': None, #lambda data: hashlib.sha3_512(data).hexdigest(),
+        'PSA_ALG_SHAKE256_512': None, #lambda data: hashlib.shake_256(data).hexdigest(64),
+    } #typing: Optional[Dict[str, Callable[[bytes], str]]]
```
Indeed, and mypy would have complained if I'd written an actual type annotation, which needs `type:`, not `typing:`.

tgonzalezorlandoarm replied:
Great, everything else is okay on my side :)

davidhorstmann-arm reviewed (dismissed):
LGTM, thanks!

davidhorstmann-arm commented:
This needs a rebase, but I'll be on-hand to re-review quickly once that is done

gilles-peskine-arm commented:
I rebased on top of the latest `development` to address the conflict between "Move PSA information and dependency automation into their own module" and changes to the moved code in `development`.

davidhorstmann-arm reviewed (dismissed):
LGTM

tgonzalezorlandoarm commented:
I think the CI is pointing to a missing import

tgonzalezorlandoarm commented on scripts/mbedtls_dev/crypto_data_tests.py:
```diff
@@ -0,0 +1,123 @@
+        'PSA_ALG_SHA3_384': None, #lambda data: hashlib.sha3_384(data).hexdigest(),
+        'PSA_ALG_SHA3_512': None, #lambda data: hashlib.sha3_512(data).hexdigest(),
+        'PSA_ALG_SHAKE256_512': None, #lambda data: hashlib.shake_256(data).hexdigest(64),
+    } #type: Dict[str, Optional[Callable[[bytes], str]]]
```
I'm expecting this to happen here (extrapolating from an error in my own PR):
What type is this dictionary? Well, I think it's actually not `Dict[str, Optional[Callable[[bytes], str]]]` and that's because lambda type is actually inferred and cannot be specified before hand, please see [this](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#callable-types-and-lambdas) for more information. 
I'm not 100% sure that this will happen so please take it with a pinch of salt.

gilles-peskine-arm commented:
I rebased again to add the missing `import`. In my previous rebase (https://github.com/Mbed-TLS/mbedtls/commit/cf22c41d29c433700e9f8cc568c4a90378184740) I'd noticed that the commit "Move PSA information and dependency automation into their own module" added `import crypto_data_tests` even though `crypto_data_tests.py` didn't exist yet, so I removed the line that added the import, but then I forgot to add it to the commit that adds the module. So I've fixed this now.

davidhorstmann-arm reviewed (approved):
LGTM, based on diff from rebase

tgonzalezorlandoarm reviewed (approved):
LGTM
